File: go/exercises/practice/unit_test_eb3d70f7_63d1_439c_beed_b31b03253e69/quantum_routing/quantum_routing_test.go
package quantum

import (
	"reflect"
	"testing"
)

func TestFindMostReliablePath(t *testing.T) {
	tests := []struct {
		name        string
		n           int
		graph       [][]Edge
		source      int
		destination int
		want        []int
	}{
		{
			name:        "Example case",
			n:           4,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.9}, {To: 2, Fidelity: 0.8}}, {{To: 0, Fidelity: 0.9}, {To: 3, Fidelity: 0.7}}, {{To: 0, Fidelity: 0.8}, {To: 3, Fidelity: 0.6}}, {{To: 1, Fidelity: 0.7}, {To: 2, Fidelity: 0.6}}},
			source:      0,
			destination: 3,
			want:        []int{0, 1, 3}, // This is one possible optimal solution
		},
		{
			name:        "Single node, source=destination",
			n:           1,
			graph:       [][]Edge{{}},
			source:      0,
			destination: 0,
			want:        []int{0},
		},
		{
			name:        "Two nodes with direct connection",
			n:           2,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.8}}, {{To: 0, Fidelity: 0.8}}},
			source:      0,
			destination: 1,
			want:        []int{0, 1},
		},
		{
			name:        "Path via intermediate node is more reliable",
			n:           3,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.5}, {To: 2, Fidelity: 0.3}}, {{To: 0, Fidelity: 0.5}, {To: 2, Fidelity: 0.9}}, {{To: 0, Fidelity: 0.3}, {To: 1, Fidelity: 0.9}}},
			source:      0,
			destination: 2,
			want:        []int{0, 1, 2}, // Path 0->1->2 has reliability 0.5*0.9=0.45, better than direct 0->2 with 0.3
		},
		{
			name:        "Disconnected graph",
			n:           4,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.7}}, {{To: 0, Fidelity: 0.7}}, {{To: 3, Fidelity: 0.8}}, {{To: 2, Fidelity: 0.8}}},
			source:      0,
			destination: 3,
			want:        []int{}, // No path exists
		},
		{
			name:        "Complex network with multiple possible paths",
			n:           5,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.9}, {To: 2, Fidelity: 0.7}}, {{To: 0, Fidelity: 0.9}, {To: 2, Fidelity: 0.8}, {To: 3, Fidelity: 0.6}}, {{To: 0, Fidelity: 0.7}, {To: 1, Fidelity: 0.8}, {To: 3, Fidelity: 0.7}, {To: 4, Fidelity: 0.5}}, {{To: 1, Fidelity: 0.6}, {To: 2, Fidelity: 0.7}, {To: 4, Fidelity: 0.8}}, {{To: 2, Fidelity: 0.5}, {To: 3, Fidelity: 0.8}}},
			source:      0,
			destination: 4,
			want:        []int{0, 2, 4}, // Path 0->2->4 has reliability 0.7*0.5=0.35
		},
		{
			name:        "Linear network",
			n:           5,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.9}}, {{To: 0, Fidelity: 0.9}, {To: 2, Fidelity: 0.8}}, {{To: 1, Fidelity: 0.8}, {To: 3, Fidelity: 0.7}}, {{To: 2, Fidelity: 0.7}, {To: 4, Fidelity: 0.6}}, {{To: 3, Fidelity: 0.6}}},
			source:      0,
			destination: 4,
			want:        []int{0, 1, 2, 3, 4},
		},
		{
			name:        "Star network with center as source",
			n:           5,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.9}, {To: 2, Fidelity: 0.8}, {To: 3, Fidelity: 0.7}, {To: 4, Fidelity: 0.6}}, {{To: 0, Fidelity: 0.9}}, {{To: 0, Fidelity: 0.8}}, {{To: 0, Fidelity: 0.7}}, {{To: 0, Fidelity: 0.6}}},
			source:      0,
			destination: 4,
			want:        []int{0, 4},
		},
		{
			name:        "Full graph with varying fidelities",
			n:           4,
			graph:       [][]Edge{{{To: 1, Fidelity: 0.5}, {To: 2, Fidelity: 0.6}, {To: 3, Fidelity: 0.3}}, {{To: 0, Fidelity: 0.5}, {To: 2, Fidelity: 0.7}, {To: 3, Fidelity: 0.4}}, {{To: 0, Fidelity: 0.6}, {To: 1, Fidelity: 0.7}, {To: 3, Fidelity: 0.8}}, {{To: 0, Fidelity: 0.3}, {To: 1, Fidelity: 0.4}, {To: 2, Fidelity: 0.8}}},
			source:      0,
			destination: 3,
			want:        []int{0, 2, 3}, // Path 0->2->3 has reliability 0.6*0.8=0.48, better than direct 0->3 with 0.3
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := FindMostReliablePath(tt.n, tt.graph, tt.source, tt.destination)
			if len(tt.want) == 0 {
				if len(got) != 0 {
					t.Errorf("FindMostReliablePath() = %v, want empty path for disconnected graph", got)
				}
				return
			}
			
			// If there's a valid path, we need to check if it's correct
			// We can check that we get from source to destination
			if got[0] != tt.source || got[len(got)-1] != tt.destination {
				t.Errorf("FindMostReliablePath() path doesn't start at source and end at destination; got = %v", got)
				return
			}
			
			// Verify the path is valid by checking each step
			for i := 0; i < len(got)-1; i++ {
				current := got[i]
				next := got[i+1]
				edgeExists := false
				for _, edge := range tt.graph[current] {
					if edge.To == next {
						edgeExists = true
						break
					}
				}
				if !edgeExists {
					t.Errorf("FindMostReliablePath() returned invalid path %v; no edge between %d and %d", got, current, next)
					return
				}
			}
			
			// If the expected path is one of several with optimal reliability, we can skip exact path comparison
			// However, if we're given a specific expected path, check that one exactly
			if tt.name == "Example case" || reflect.DeepEqual(got, tt.want) {
				// Calculate reliability of returned path
				reliability := calculatePathReliability(tt.graph, got)
				expectedReliability := calculatePathReliability(tt.graph, tt.want)
				
				// Allow for slight floating point differences
				const epsilon = 1e-9
				if reliability < expectedReliability-epsilon {
					t.Errorf("FindMostReliablePath() returned path with reliability %v, want %v", reliability, expectedReliability)
				}
			}
		})
	}
}

// Helper function to calculate path reliability
func calculatePathReliability(graph [][]Edge, path []int) float64 {
	if len(path) <= 1 {
		return 1.0 // A path with one or zero nodes has perfect reliability
	}
	
	reliability := 1.0
	for i := 0; i < len(path)-1; i++ {
		current := path[i]
		next := path[i+1]
		
		// Find the edge connecting current and next
		for _, edge := range graph[current] {
			if edge.To == next {
				reliability *= edge.Fidelity
				break
			}
		}
	}
	
	return reliability
}

func BenchmarkFindMostReliablePath(b *testing.B) {
	// Create a reasonably sized graph for benchmarking
	n := 100
	graph := make([][]Edge, n)
	for i := 0; i < n; i++ {
		// Connect to 5 random nodes
		graph[i] = []Edge{}
		for j := 1; j <= 5; j++ {
			next := (i + j) % n
			graph[i] = append(graph[i], Edge{To: next, Fidelity: 0.8})
		}
	}
	
	source := 0
	destination := n/2
	
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		FindMostReliablePath(n, graph, source, destination)
	}
}

File: cpp/exercises/practice/unit_test_39947b5a_fd9f_468a_abc9_2b7eaf7f9430/quantum_routing/quantum_routing_test.cpp
#include "quantum_routing.h"
#include "catch.hpp"
#include <cmath>
#include <vector>

bool is_close(double a, double b, double epsilon = 1e-9) {
    return std::abs(a - b) < epsilon;
}

TEST_CASE("Same source and destination", "[quantum_routing]") {
    int n = 5;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    
    int start = 3;
    int end = 3;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 1.0));
}

TEST_CASE("Direct connection", "[quantum_routing]") {
    int n = 5;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 0.5;
    
    int start = 0;
    int end = 1;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.5));
}

TEST_CASE("No path exists", "[quantum_routing]") {
    int n = 5;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    
    int start = 0;
    int end = 4;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.0));
}

TEST_CASE("Indirect connection with better probability", "[quantum_routing]") {
    int n = 3;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 0.5;
    probabilities[1][2] = 0.5;
    probabilities[0][2] = 0.1;
    
    int start = 0;
    int end = 2;
    
    // Direct: 0->2 with 0.1 probability
    // Indirect: 0->1->2 with 0.5*0.5 = 0.25 probability
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.25));
}

TEST_CASE("Complex network", "[quantum_routing]") {
    int n = 5;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 0.9;
    probabilities[0][2] = 0.8;
    probabilities[1][2] = 0.7;
    probabilities[1][3] = 0.6;
    probabilities[2][3] = 0.5;
    probabilities[2][4] = 0.4;
    probabilities[3][4] = 0.9;
    
    int start = 0;
    int end = 4;
    
    // Path options:
    // 0->2->4 with 0.8*0.4 = 0.32 probability
    // 0->1->3->4 with 0.9*0.6*0.9 = 0.486 probability
    // 0->2->3->4 with 0.8*0.5*0.9 = 0.36 probability
    // 0->1->2->4 with 0.9*0.7*0.4 = 0.252 probability
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.486));
}

TEST_CASE("Asymmetric probabilities", "[quantum_routing]") {
    int n = 4;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 0.8;
    probabilities[1][0] = 0.5; // Different from 0->1
    probabilities[1][2] = 0.7;
    probabilities[2][1] = 0.4; // Different from 1->2
    probabilities[2][3] = 0.6;
    probabilities[3][2] = 0.3; // Different from 2->3
    
    int start = 0;
    int end = 3;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.8 * 0.7 * 0.6));
}

TEST_CASE("Multiple paths", "[quantum_routing]") {
    int n = 7;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    // Path 1: 0->1->3->6
    probabilities[0][1] = 0.9;
    probabilities[1][3] = 0.8;
    probabilities[3][6] = 0.7;
    
    // Path 2: 0->2->4->6
    probabilities[0][2] = 0.8;
    probabilities[2][4] = 0.9;
    probabilities[4][6] = 0.8;
    
    // Path 3: 0->5->6
    probabilities[0][5] = 0.75;
    probabilities[5][6] = 0.85;
    
    int start = 0;
    int end = 6;
    
    // Path 1: 0.9*0.8*0.7 = 0.504
    // Path 2: 0.8*0.9*0.8 = 0.576
    // Path 3: 0.75*0.85 = 0.6375
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.6375));
}

TEST_CASE("Cycle in the network", "[quantum_routing]") {
    int n = 4;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 0.9;
    probabilities[1][2] = 0.8;
    probabilities[2][0] = 0.7; // Creates a cycle
    probabilities[1][3] = 0.6;
    
    int start = 0;
    int end = 3;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.9 * 0.6));
}

TEST_CASE("Large network", "[quantum_routing]") {
    int n = 100;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    
    // Create a chain: 0->1->2->...->99
    for (int i = 0; i < n - 1; i++) {
        probabilities[i][i+1] = 0.99;
    }
    
    int start = 0;
    int end = n - 1;
    
    // Expected probability: 0.99^99
    double expected = std::pow(0.99, 99);
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), expected));
}

TEST_CASE("Network with disconnected components", "[quantum_routing]") {
    int n = 10;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    
    // Component 1: 0, 1, 2, 3, 4
    probabilities[0][1] = 0.9;
    probabilities[1][2] = 0.8;
    probabilities[2][3] = 0.7;
    probabilities[3][4] = 0.6;
    
    // Component 2: 5, 6, 7, 8, 9
    probabilities[5][6] = 0.9;
    probabilities[6][7] = 0.8;
    probabilities[7][8] = 0.7;
    probabilities[8][9] = 0.6;
    
    // No path from component 1 to component 2
    int start = 0;
    int end = 9;
    
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), 0.0));
}

TEST_CASE("Edge case: very small probabilities", "[quantum_routing]") {
    int n = 5;
    std::vector<std::vector<double>> probabilities(n, std::vector<double>(n, 0.0));
    probabilities[0][1] = 1e-9;
    probabilities[1][2] = 1e-9;
    probabilities[2][3] = 1e-9;
    probabilities[3][4] = 1e-9;
    
    int start = 0;
    int end = 4;
    
    double expected = 1e-36;
    // Use a larger epsilon for very small numbers
    REQUIRE(is_close(quantum_routing::find_highest_probability(probabilities, start, end), expected, 1e-40));
}